### PR TITLE
automate checking out pull request

### PR DESCRIPTION
**Pull request by kendonB**
This pull request automates a very common task: checking out a pull request from GitHub.

The user needs to have a file in `.caster/data` called `local_remote_git_match.toml` of the form:

```
[github]
"https://github.com/dictation-toolbox/Caster" = "%USERPROFILE%\\Documents\\caster"
``` 

The user just has to say "checkout pull request" while at the pull request in the browser.  caster will take over and do all the boilerplate stuff, then the user just has to press "enter" at the end to confirm the creation of the branch.

To test this, you will need to replace the `C:/Users/kmbel/.caster/data/local_remote_git_match.toml` in `automation.py` with your own file location. This still needs a location in settings.toml.

This begins to address #494.

- [ ] @kendonB test it is working after all changes
- [x] Add a setting for the location of the local_remote_git_match.toml file. Currently it just refers to `C:/Users/kmbel/.caster/data/local_remote_git_match.toml` so that's not going to work for the next tester until they change that.
- [x] Refactor the actual call out of `nav.py` into where? `auto.py`? what folder?
- [x] Consider a narrower name like "Github automation"
- [ ] Someone else test it is working
- [x] Add a setting for the git bash location in settings.toml
- [x] Remove debugging commands
- [x] Add a blank toml file with a commented out example
- [x] Add docs

### CODE REVIEW

**Comment by LexiconCode** [comment]
I will take a look at this next week after finals.

**Comment by kendonB** [comment]
I won't touch the rebuild items function for now as it will be cleaner to move those both into `utilities.py` in another pull request

**Comment by LexiconCode** [comment]
In the next few days I will try to review this and test.

**Comment by LexiconCode** [comment]
I gotta say this is really neat!

a bit of testing with  "check out pull request"

the status window
```
Github: Checkout pull request locally
Checking out pull request locally.
Checking out pull request locally 2.
https://github.com/dictation-toolbox/Caster/pull/490
```

However when I view the repository
`user@Main-Desk MINGW64 /d/Backup/Library/Documents/Caster (pr/n498_automation)`

local_remote_git_match.toml is blank

**Comment by kendonB** [comment]
> However when I view the repository
user@Main-Desk MINGW64 /d/Backup/Library/Documents/Caster (pr/n498_automation)

> local_remote_git_match.toml is blank

The user needs to create the association between the remote URLs and the local repo locations in that file. Here is mine. To be clear, you see `local_remote_git_match.toml` but it's empty?
```
[github]
"https://github.com/dictation-toolbox/Caster" = "%USERPROFILE%\\Documents\\caster"
```

**Comment by LexiconCode** [comment]
> 
> 
> > However when I view the repository
> > user@Main-Desk MINGW64 /d/Backup/Library/Documents/Caster (pr/n498_automation)
> 
> > local_remote_git_match.toml is blank
> 
> The user needs to create the association between the remote URLs and the local repo locations in that file. Here is mine. To be clear, you see `local_remote_git_match.toml` but it's empty?
> 
> ```
> [github]
> "https://github.com/dictation-toolbox/Caster" = "%USERPROFILE%\\Documents\\caster"
> ```

Yes it is empty.

There might be an easier way. Why not read the information you need from ".git"? For example 'config'

File: castervoice/lib/settings.py [hunk 1]
```diff
@@ -168,7 +172,11 @@ def _find_natspeak():
 
         # PYTHON
         "PYTHONW": "C:/Python27/pythonw",
-        "WXPYTHON_PATH": "C:/Python27/Lib/site-packages/wx-3.0-msw"
+        "WXPYTHON_PATH": "C:/Python27/Lib/site-packages/wx-3.0-msw",
+        "TERMINAL_PATH": TERMINAL_PATH_DEFAULT,
+
```
Comment (lexxish) [inline castervoice/lib/settings.py]: new lines?

File: castervoice/lib/automation.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        import natlink
+        context = AppContext(executable="chrome") | AppContext(executable="iexplore") | AppContext(executable="firefox")
+        if context:
+            print("Checking out pull request locally.")
+            Key("c-l/20").execute()
+            print("Checking out pull request locally 2.")
+            url = read_selected_without_altering_clipboard()
+            print(url[1])
+            if url[0] == 0:
+                split_string = url[1].split("/pull/")
```
Comment (lexxish) [inline castervoice/lib/automation.py]: Can this be named better

**Comment by lexxish** [comment]
Not sure why I'm commenting when it's not done. Based on the description, shouldn't this be in the apps specific grammars?

**Comment by kendonB** [comment]
> There might be an easier way. Why not read the information you need from ".git"? For example 'config'

All we know upon invocation of the command is the URL. We don't know where the .git folder is. Unless I'm missing something

File: castervoice/lib/automation.py (hunk 2 above)
Comment (kendonB) [reply castervoice/lib/automation.py]: Which name?

File: castervoice/lib/settings.py (hunk 1 above)
Comment (kendonB) [reply castervoice/lib/settings.py]: will remove. Thanks

**Comment by kendonB** [comment]
> Not sure why I'm commenting on the code when it's not done. Based on the description, shouldn't this be in the apps specific grammars?

@lexxish reviews are welcome at any stage.

**Comment by lexxish** [comment]
I mean under the apps section for chrome/firefox

File: castervoice/lib/automation.py (hunk 2 above)
Comment (lexxish) [reply castervoice/lib/automation.py]: Yes. Maybe split the array into two variables with names

**Comment by kendonB** [comment]
> I mean under the apps section for chrome/firefox

I agree with this. Might be worth finishing #530 first then and we can add this to the standard browser grammar.

**Comment by LexiconCode** [comment]
> 
> 
> > There might be an easier way. Why not read the information you need from ".git"? For example 'config'
> 
> All we know upon invocation of the command is the URL. We don't know where the .git folder is. Unless I'm missing something

For the classic install this is very predictable as it always be in the caster root folder. Pip. I have a (rudimentary) function that checks whether or not it's a classic or PIP install. Since this is for caster development the classic install would have to be used.
https://github.com/dictation-toolbox/Caster/blob/22661329acbfb543d6264446d75a2f1ec0f07dab/_caster.py#L69

**Comment by kendonB** [comment]
> Since this is for caster development the classic install would have to be used.

This should work for an arbitrary GitHub repo, not just caster. That's the intention. I'm imagining users might have half a dozen repos they're interested in checking out the pull requests of.

**Comment by LexiconCode** [comment]
Okay excellent!

**Comment by lexxish** [comment]
> > Since this is for caster development the classic install would have to be used.
> 
> This should work for an arbitrary GitHub repo, not just caster. That's the intention. I'm imagining users might have half a dozen repos they're interested in checking out the pull requests of.

Will it also be possible to check out a branch? For example https://github.com/dictation-toolbox/Caster/tree/develop. I'm assuming there will be a setting somewhere for "PROJECT_WORKSPACE" that we will point to where we keep our code.

**Comment by kendonB** [comment]
Checking out branches from the browser should also be possible but we'll do that in a separate PR. I'm not as sure of the use case for that though as mostly I'm updating branches using git pull not making fresh versions

**Comment by kendonB** [comment]
I think this is ready to go. Can someone else test? @LexiconCode or @lexxish?

**Comment by LexiconCode** [comment]
I might be able to win the weekend comes but during the week I'm pretty booked.

**Comment by codingApprentice** [comment]
@kendonB I tested code and it seems to be working fine for me. I just needed to increase some of the time delays and combine some of the commands to ensure they were performed in sequence.

- git bash would take just under 5 seconds to initially load

- Fetch commands needed to be given time to complete

I also changed it so that you can perform the pull request voice command on any of the pull request page tabs, not just the conversation one (so it should work on commits, checks, and file changes tabs). See [my pull request on this pull request ](https://github.com/kendonB/Caster/pull/2)

Let me know what you think. :)

**Comment by kendonB** [comment]
Asked https://stackoverflow.com/questions/56486172/how-do-i-get-popen-to-wait-for-git-bash-exe-to-be-ready-for-keyboard-input

**Comment by codingApprentice** [comment]
[Update using pywinauto instead of popen](https://github.com/kendonB/Caster/pull/2#issuecomment-500067305)

- Updated to use pywinauto (instead of popen) which waits until the process is idle after initialisation before inputting text.

- pywinauto can apparently have issues when interacting between 32 bit Python and 64-bit binaries (git bash), but there were no observed issues with my change

- I included a check to use 32 bit git-bash when using 32 bit Python (this would of course will require the user to install 32 bit git-bash as well as the usual 64-bit version)

- Despite this check, it still seems to launch the 64-bit version of git-bash. I'm not sure why. Nevertheless, I just seem to get the warning message about mixing different versions, but no errors in execution.

**Comment by kendonB** [comment]
@codingApprentice is the below warning a concern? If not, can we suppress it?

```
C:\Python27\lib\site-packages\pywinauto\__init__.py:80: UserWarning: Revert to STA COM threading mode
  warnings.warn("Revert to STA COM threading mode", UserWarning)
```

**Comment by kendonB** [comment]
I presume this is the warning you see?
```
C:\Python27\lib\site-packages\pywinauto\application.py:997: UserWarning: 64-bit binary from 32-bit Python may work incorrectly (please use 64-bit Python instead)
  _warn_incorrect_binary_bitness(exe_name)
C:\Python27\lib\site-packages\pywinauto\application.py:1068: UserWarning: 64-bit application should be automated using 64-bit Python (you use 32-bit Python)
  UserWarning)
C:\Python27\lib\site-packages\pywinauto\application.py:1053: RuntimeWarning: Application is not loaded correctly (WaitForInputIdle failed)
  warnings.warn('Application is not loaded correctly (WaitForInputIdle failed)', RuntimeWarning)
```
On mine, it still waits a good number of seconds after opening before doing anything.

**Comment by kendonB** [comment]
Your other changes will seem to be working great. Thanks!

**Comment by codingApprentice** [comment]
[Switch from pywinauto to autohotkey to fix errors and speed up execution #3](https://github.com/kendonB/Caster/pull/3#issue-286504018)
Rethinking my previous pull request, I managed to find a faster way of executing.

- The user just needs to install the latest version of [autohotkey](https://autohotkey.com/download/1.1/)

- This new approach also considers whether an existing git bash windows open already or not

- It should reduce the execution time from several seconds to a single second once the git bash window is open

**Comment by LexiconCode** [comment]
I'm going to remove the remove 'needs review' as reviewing is in progress with multiple contributors. I can always add it back if there's other thoughts on the matter.

**Comment by kendonB** [comment]
@codingApprentice I don't think that the waiting issue is such a big deal as to warrant an extra dependency (autohotkey). Is it possible to pull out whatever they do and put the source into caster?

**Comment by codingApprentice** [comment]
@kendonB 
- If dependencies are a concern, then perhaps it's best to revert back to:
    - either pywinauto, and accept the warning messages until the switch to 64-bit Python when the switch to Python 3 is completed
    - or back to the simple time.sleep(5)… Which was how long it was taking git bash to load, for me at least.

- I did see an [AHK pip library](https://pypi.org/project/ahk/), but it requires Python 3.

- Do you mean putting the source into caster as in, include the executable for autohotkey in the caster repository itself? Or do you mean something else?

**Comment by kendonB** [comment]
> Do you mean putting the source into caster as in, include the executable for autohotkey in the caster repository itself? Or do you mean something else?

I meant just pulling out the AHK source code that's relevant to checking the status of windows. Doing this will obviously be a bunch of work if AHK is not written in python. If not, we would compile to a dll (I'm sure there's a nice pythonic way to do this for a bunch of languages). This functionality can also improve the "bringme" experience so I think it wouldn't be a waste of time. That being said, we will all be on Python 3 within 6 months and adding another pip package is less of a worry.

I tried it with AHK and it's quite nice. In the meantime, can we switch to detecting if the user has AHK? And let's also:

- [ ] Add AHK to the caster installation page as an optional extra
- [ ] Add the pause time to the settings toml as a variable
- [ ] go through old issues to see where AHK can help

**Comment by kendonB** [comment]
@codingApprentice I made a few final tweaks to your latest version:
 - I allowed for `__file__` to also be a `.pyc` file.
 - A minor change to the documentation
 - If an existing Git Bash session gets used, it will use `end, c-u` to delete everything that might be sitting there in the command bit of the terminal.

I'm going to merge now but please make sure it still works for you!

**Comment by kendonB** [comment]
@codingApprentice many thanks for your contributions to this!

**Comment by codingApprentice** [comment]
@kendonB yes it still works for me. And today I learned the ctrl + u shortcut for git bash.

**Pull request merged**